src/btree/col_modify.c

```diff
@@ -15,7 +15,7 @@ static int __col_insert_alloc(WT_SESSION_IMPL *, uint64_t, u_int, WT_INSERT **,
  *     Column-store delete, insert, and update.
  */
 int
-__wt_col_modify(WT_CURSOR_BTREE *cbt, uint64_t recno, const WT_ITEM *value, WT_UPDATE *upd_arg,
+__wt_col_modify(WT_CURSOR_BTREE *cbt, uint64_t recno, const WT_ITEM *value, WT_UPDATE **updp_arg,
```

Comment:
Can we change the code to always free the memory in failure in the caller? I think it should be the caller to __wt_update_serial that allocates the memory and it should be the caller to manage the life cycle of the memory.

Reply:
Why would this be advantageous? We could change `__wt_update_serial` so that the caller has to free the memory on error if we'd like, but this wouldn't solve the problem, as the issue is with the memory management by `__wt_col_modify` and `__wt_row_modify`. These functions are fairly complicated, and they are called from many places, so I would prefer to keep the memory management to what I proposed in the PR, as that's a relatively minimally invasive change.

But I do agree with you that we could do a better and a more consistent job with memory management. That would be probably a larger change that requires a bit more thought and would be better done in a separate ticket. In that case, we would not have to rush with that change, as this is an 8.0 required ticket.

Reply:
understood